APIKit/Request.swift

```diff
@@ -1,9 +1,95 @@
 import Foundation
+import Result
 
+/// Request protocol represents a request for Web API.
+/// Following 5 items must be implemented.
+/// - typealias Response
+/// - var baseURL: NSURL
+/// - var method: Method
+/// - var path: String
+/// - func responseFromObject(object: AnyObject, URLResponse: NSHTTPURLResponse) throws -> Response
 public protocol Request {
-    typealias Response: Any
-    
-    var URLRequest: NSURLRequest? { get }
-    
-    static func responseFromObject(object: AnyObject) -> Response?
+    /// Type represents a model object
+    typealias Response
+
+    /// Configurations of request
+    var baseURL: NSURL { get }
+    var method: HTTPMethod { get }
+    var path: String { get }
+    var parameters: [String: AnyObject] { get }
+
+    /// You can add any configurations here
+    func configureURLRequest(URLRequest: NSMutableURLRequest) throws -> NSMutableURLRequest
+
+    /// Set of status code that indicates success.
+    /// `responseFromObject(_:URLResponse:)` will be called if this contains NSHTTPURLResponse.statusCode.
+    /// Otherwise, `errorFromObject(_:URLResponse:)` will be called.
+    var acceptableStatusCodes: Set<Int> { get }
+
+    /// An object that builds body of HTTP request.
+    var requestBodyBuilder: RequestBodyBuilder { get }
+
+    /// An object that parses body of HTTP response.
+    var responseBodyParser: ResponseBodyParser { get }
+
+    /// Build `Response` instance from raw response object.
+    /// This method will be called if `acceptableStatusCode` contains status code of NSHTTPURLResponse.
+    func responseFromObject(object: AnyObject, URLResponse: NSHTTPURLResponse) throws -> Response
+
+    /// Build `ErrorType` instance from raw response object.
+    /// This method will be called if `acceptableStatusCode` does not contain status code of NSHTTPURLResponse.
+    func errorFromObject(object: AnyObject, URLResponse: NSHTTPURLResponse) throws -> ErrorType
```

Comment:
I understand your intention. But `.UnexpectedResponse` lacks meaningful info about _what is unexpected_. The caller can't detect the root cause of that error without debugging and stack traces. So I think there are two points we could address here:

1.  Whether notifying "unexpected response" from `errorFromObject()` is implementor's responsibility or not. They may have a following error type and throw it in `errorFromObject()`:

    ```swift
    enum ErrorResponse: ErrorType {
        case InvalidRawResponse(object: AnyObject, URLResponse: NSHTTPURLResponse)
        case BadRequest(message: String)
    }
    ```

2. Add an/some associated values to `.UnexpectedResponse`. Is it `ErrorType` or `NSURLResponse?` or `(object: AnyObject, URLResponse: NSHTTPURLResponse)`?

How do you think about my thought above?